tokio/src/sync/rwlock.rs

```diff
@@ -375,8 +375,6 @@ impl<'a, T: ?Sized> RwLockWriteGuard<'a, T> {
     /// let n = n.downgrade();
     /// assert_eq!(*n, 1, "downgrade is atomic");
     ///
-    /// assert_eq!(*lock.read().await, 1, "additional readers can obtain locks");
```

Comment:
I don't understand how this would hang. If all permits as assigned to the spawned task, wouldn't they be released once the write lock is dropped?

Reply:
You are right, I was getting confused by the loom test in which it blocks

Reply:
@carllerche I took another look at all that and I think it can actually hang. I might be way off here but consider the following. If we look at the loom version of the test: 

```
fn downgrade() {
    tracing_subscriber::fmt::init();
    loom::model(|| {
        let lock = Arc::new(RwLock::new(1));

        let n = block_on(lock.write().instrument(tracing::info_span!("write_1")));
        let cloned_lock = lock.clone();
        let handle = thread::spawn(move || {
            let mut guard = block_on(cloned_lock.write() .instrument(tracing::info_span!("write_2")));
            *guard = 2;
        });

        let span = tracing::span!(tracing::Level::TRACE, "downgrading");
        let enter = span.enter();

        let n = n.downgrade();
        assert_eq!(*n, 1);
        drop(enter);

        assert_eq!(*block_on(lock.read().instrument(tracing::info_span!("read_1"))), 1);

        drop(n);

        handle.join().unwrap();
        assert_eq!(*block_on(lock.read() .instrument(tracing::info_span!("read_2"))), 2);
    });
}
```
So consider the following sequence of events:

- `write_1` locks taking 10 permits and leaving 0
- `write_2` is spawned trying to take 10 permits and getting queued in the waitlist
- `write_1` is downgraded, releasing 9 permits and leaving one for the read handle that it produces
- `write_2` is woken up and acquires all 9 available permits, leaving 0
-  at this point we have one read lock that holds one permit and our waiter's queue has the `write_2` waiter holding 9 permits and still needing one
- we block on `read_1` trying to acquire a read lock, requesting 1 permit.
- at that point we hang... because noone can make progress
- `write_2` needs 1 permit so it will not proceed
- `read_1` needs one permit so it will not proceed
- `n` holds this permit but it releases it only after `read_1` is done

Reply:
Ah, of course, because it is a fair RwLock...